Create /register and /logout endpoints. Fixes #11 && second half of #17

khaldarov opened the pull request:
Endpoint does register to the DB. NOT ready for merge. Requesting review for the user model and server-side validation.

- Added `express-validator` and changed its version.
- Changed user model to make it suitable with 3rd party authentication.

khaldarov committed:
Send to review

khaldarov commented:
Validator does work, but not as intended: it does not show the custom error messages.
![Screenshot from 2022-07-27 23-36-05](https://user-images.githubusercontent.com/102853969/181584038-9f243858-81b7-4e63-a5ef-4fd639dd9794.png)

khatibAmjad commented:
The username can still be required, I was able to create a username from the email by deleting `@gmail.com` from the email. Other than that the changes you made are enough for the 3rd party authentication.

As for the validator I didn't get the problem, the custom message you wrote is invalid email and that's the message the error gave you on postman. Again maybe I didn't get the question.

One more thing, I think all the server side validation should be done in a middleware not in the controller.

khatibAmjad commented:
For the user schema we should also remove the `unique: true` from the fields that aren't required, coz i faced an error now telling me that i already have a null phone number ana i can't register another user with the same value

khaldarov commented:
I applied asked revisions. I also have 2 questions:
- Wouldn't using only the first half of Gmail address as username intervene with existing users? Like I might have a username `adnan`, and when some other user tries to register with the email `adnan@gmail.com`  he will get an error.
- Would removing the `uniqe: true` field be a good approach?
Now I'm thinking of what would've it been like if we had used a different model for 3rd party folks.

Other than that, I didn't use `express-validator`, since it didn't work as I intended.
I will do the validation in a middleware, noted.

khaldarov committed:
Fix requested revision, add validation middleware

khaldarov committed:
Fix validation middleware

khaldarov committed:
Create /register and /logout endpoints.
Fixes #11 and second half of #17

khaldarov committed:
Fix test exit issue && add `disconnect` after test

khaldarov commented:
- Finished _/register_ and _/logout_ endpoints; added GET _/register_ render endpoint.
- Added tests for different cases: good path, bad path, and horrible path (lol).
- Fixed server-side validation, making all errors elements of an `errorsArray`.
- Changed user model according to our new needs.
- Added documentation and 3 examples for the _/register_ endpoint; added schema field in documentation.

khatibAmjad commented on src/controllers/auth.js:
```diff
@@ -1,9 +1,13 @@
-// const jwt = require('jsonwebtoken');
 const UserModel = require('../models/user');
 
+const registerPage = (_, res) => {
```
Isn't this frontend responsibility? I don't think we need to render anything.

khaldarov replied:
Sure thing. I can remove that.

khatibAmjad commented on src/models/user.js:
```diff
@@ -135,6 +131,7 @@ const userSchema = new mongoose.Schema(
     },
     avaregeRating: {
       type: Number,
+      required: true,
```
I think this shouldn't be required

khaldarov replied:
All the ones that have a `default` will not be a problem when using 3rd party auth. No need to specify them; they'll be added automatically, by default. I wanted to make them look unified. I still can change them if they're causing any issues.

khatibAmjad replied:
I vote for removing the required since they aren't used in this case.

khaldarov replied:
All the `required: true` fields that also have a `default` are only to make the model look unified; they can be removed. On the other hand, they do not really affect how the database functions.

I thought of the profile photo causing a problem when is removed, but there _will_ be a default, placeholder, image for the profile photo. I still can remove the `required: true` fields if you think they'll cause any issues.

khaldarov replied:
Removed them 👍🏽

khatibAmjad commented on src/models/user.js:
```diff
@@ -147,16 +144,19 @@ const userSchema = new mongoose.Schema(
     donated: {
       ref: 'Donation',
       type: [objectId],
+      required: true,
```
also the required here will be a problem for the 3rd party authentication

khatibAmjad commented on src/models/user.js:
```diff
@@ -147,16 +144,19 @@ const userSchema = new mongoose.Schema(
     requested: {
       ref: 'Request',
       type: [objectId],
+      required: true,
```
same here

khatibAmjad commented on src/models/user.js:
```diff
@@ -147,16 +144,19 @@ const userSchema = new mongoose.Schema(
     received: {
       ref: 'Request',
       type: [objectId],
+      required: true,
```
same here

khatibAmjad commented on src/models/user.js:
```diff
@@ -174,14 +174,17 @@ const userSchema = new mongoose.Schema(
     },
     address: {
       type: [addressSchema],
+      required: true,
```
same here

khatibAmjad commented on src/models/user.js:
```diff
@@ -174,14 +174,17 @@ const userSchema = new mongoose.Schema(
     },
     reviews: {
       type: [reviewSchema],
+      required: true,
```
same here

khatibAmjad commented on src/models/user.js:
```diff
@@ -174,14 +174,17 @@ const userSchema = new mongoose.Schema(
     },
     reports: {
       type: [reportSchema],
+      required: true,
```
same here

khatibAmjad commented on src/docs/doc.yaml:
```diff
@@ -36,19 +30,116 @@ paths:
                 password1:
                   type: string,
+            examples:
+              'SLIGHTLY BAD PATH':
```
what's the point of having two different bad entries? aren't they going to give the same result?

khaldarov replied:
They return different results. One only checks the format; the other checks the database.

khatibAmjad replied:
okay

khatibAmjad commented on src/controllers/auth.js:
```diff
@@ -19,39 +23,50 @@ const register = async (req, res) => {
-        return res.status(200).redirect('/api/auth/login');
-      }
-      return res.status(401).send({ error: 'Email already exists' });
+    if (usedUsername) {
```
I suggest assigning the username to the user like we did in the 3rd party auth. That way we won't need to check the database twice for existing users.

khaldarov replied:
I didn't quite get this one. Are you suggesting auto-generated usernames, and allowing users only to choose their email addresses?

khatibAmjad replied:
exactly

khaldarov replied:
Might be a good approach. I can work on that 👍🏽

khatibAmjad reviewed (changes requested):
Impressive job! especially with the model and the regex.

khaldarov committed:
Fix requested revision && add `sparse` to model

khaldarov committed:
Fix test cases

khaldarov commented:
- Fixed requested revision.
- Added some needed constants.
- Added required documentation.
- Added `sparse` to the model for `phoneNumber` field: it now could be empty.
- Adapted test according to the new conditions.
- Changed default profile photo.

khatibAmjad commented on src/lib/constants.js:
```diff
@@ -1,4 +1,10 @@
+/* eslint-disable prefer-regex-literals */
 module.exports = Object.freeze({
+  UNIQUE_USERNAME: (email) => `${email.split('@')[0]}_${new Date().valueOf()}`,
```
I added this function to a utils fille inside the services folder like this: 
```
const generateUniqeUsername = (email) => {
  return `${email.split('@')[0]}_${new Date().valueOf()}`;
};
```
and then used it in my create method. I think it's better than putting it in the constant file.

khaldarov committed:
Fix requested revision

khaldarov committed:
Delete `console.log()`

khatibAmjad commented on src/__tests__/auth.spec.js:
```diff
@@ -1,5 +1,102 @@
+    await mongoose.disconnect();
+  });
+
+  const fakeUser = {
```
can you change this to `invalidUserInfo`

khatibAmjad commented on src/__tests__/auth.spec.js:
```diff
@@ -1,5 +1,102 @@
+    password1: 'Qwerty-123456789',
+  };
+
+  const realUser = {
```
and this to `validUserInfo`

khatibAmjad commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,62 @@
+    return res.status(201).json({ created: newUser });
+    // return res.status(201).redirect('/api/auth/login');
+  } catch (err) {
+    return res.status(500).send(err);
```
I think this should be added to the swagger

khatibAmjad commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,62 @@
+};
+
+// eslint-disable-next-line arrow-body-style
+const logout = async (req, res) => {
```
This also should have a swagger documentation

khaldarov committed:
Fix bad naming && add missing documentation

halitbatur commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,62 @@
+      errorsArray.push('Email is already taken');
+    }
+
+    if (errorsArray.length > 0) {
```
I dont see the usage of this, like before it there's only 1 error that u are checking with and its basicaly returning it as json, so I think this arr can be removed and u can have smth like this
```
if (usedEmail) {
  return res.status(400).json({error:"Email is already taken"})
}
```

khaldarov replied:
I totally forgot about that array. Now fixed.

halitbatur commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,62 @@
+      passwordHash,
+    });
+
+    return res.status(201).json({ created: newUser });
```
here lets return the newUser right away, no need to nest it insside created key

halitbatur replied:
plus I dont think its safe to return the newUser with all the data that it has with it

halitbatur commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,62 @@
+    });
+
+    return res.status(201).json({ created: newUser });
+    // return res.status(201).redirect('/api/auth/login');
```
no commented code

halitbatur commented on src/lib/constants.js:
```diff
@@ -33,4 +36,33 @@ module.exports = Object.freeze({
+  ),
+  USERNAME_REGEX: new RegExp(/^[a-zA-Z0-9\-_.]{2,20}$/),
+  PHONE_NUMBER_REGEX: new RegExp(/^[+]?[0-9]{1,3}[-\s]?[0-9]{1,3}[0-9]{4,9}$/),
+
```
extra new line

khaldarov replied:
No extra lines allowed. I had `// models' constants end` for that, but removed upon request. Added empty comment lines now instead.

halitbatur commented on src/lib/constants.js:
```diff
@@ -33,4 +36,33 @@ module.exports = Object.freeze({
   ],
   ENUM_POST_TYPE: ['Request', 'Donate'],
   ENUM_POST_STATUS: ['Published', 'Verified'],
+
```
extra new line

halitbatur replied:
@khaldarov I meant that it was an extra line, so please remove the comment and remove the extra line too

khaldarov replied:
My bad. Now fixed.

halitbatur commented on src/routes/auth.js:
```diff
@@ -1,6 +1,11 @@
 const router = express.Router();
 
+router.post('/register', registerMiddleware, authController.register);
+
```
extra new line

halitbatur reviewed (changes requested):
Good job Adnan, a few changes and it can be merged

khaldarov committed:
Fix requested revision

khaldarov committed:
Fix typo

khaldarov committed:
Fix requested revision

khaldarov committed:
Fix email-regex bug

halitbatur commented:
@khaldarov Fix conflicts and its ready to be merged

khaldarov committed:
Fix conflicts; ready to merge

khaldarov committed:
Fix conflicts; ready to merge

khaldarov committed:
Delete extra lines

khaldarov committed:
Merge branch 'dev' into khaldarov/issue11
